File: packages/agent-mesh/sdks/go/policy_test.go
package agentmesh

import (
	"os"
	"path/filepath"
	"testing"
)

func TestEvaluateExactMatch(t *testing.T) {
	pe := NewPolicyEngine([]PolicyRule{
		{Action: "data.read", Effect: Allow},
	})
	if d := pe.Evaluate("data.read", nil); d != Allow {
		t.Errorf("decision = %q, want allow", d)
	}
}

func TestEvaluateWildcard(t *testing.T) {
	pe := NewPolicyEngine([]PolicyRule{
		{Action: "data.*", Effect: Allow},
	})
	if d := pe.Evaluate("data.write", nil); d != Allow {
		t.Errorf("decision = %q, want allow", d)
	}
}

func TestEvaluateGlobalWildcard(t *testing.T) {
	pe := NewPolicyEngine([]PolicyRule{
		{Action: "*", Effect: Review},
	})
	if d := pe.Evaluate("anything", nil); d != Review {
		t.Errorf("decision = %q, want review", d)
	}
}

func TestEvaluateDefaultDeny(t *testing.T) {
	pe := NewPolicyEngine(nil)
	if d := pe.Evaluate("data.read", nil); d != Deny {
		t.Errorf("decision = %q, want deny (default)", d)
	}
}

func TestEvaluateConditions(t *testing.T) {
	pe := NewPolicyEngine([]PolicyRule{
		{Action: "data.read", Effect: Allow, Conditions: map[string]interface{}{"role": "admin"}},
	})

	if d := pe.Evaluate("data.read", map[string]interface{}{"role": "admin"}); d != Allow {
		t.Errorf("decision with matching condition = %q, want allow", d)
	}
	if d := pe.Evaluate("data.read", map[string]interface{}{"role": "guest"}); d != Deny {
		t.Errorf("decision with non-matching condition = %q, want deny", d)
	}
}

func TestEvaluateFirstMatchWins(t *testing.T) {
	pe := NewPolicyEngine([]PolicyRule{
		{Action: "data.read", Effect: Deny},
		{Action: "data.read", Effect: Allow},
	})
	if d := pe.Evaluate("data.read", nil); d != Deny {
		t.Errorf("first-match should win, got %q", d)
	}
}

func TestLoadFromYAML(t *testing.T) {
	dir := t.TempDir()
	yamlContent := `rules:
  - action: "file.read"
    effect: "allow"
  - action: "file.delete"
    effect: "deny"
`
	path := filepath.Join(dir, "policy.yaml")
	if err := os.WriteFile(path, []byte(yamlContent), 0644); err != nil {
		t.Fatal(err)
	}

	pe := NewPolicyEngine(nil)
	if err := pe.LoadFromYAML(path); err != nil {
		t.Fatalf("LoadFromYAML: %v", err)
	}

	if d := pe.Evaluate("file.read", nil); d != Allow {
		t.Errorf("YAML rule: decision = %q, want allow", d)
	}
	if d := pe.Evaluate("file.delete", nil); d != Deny {
		t.Errorf("YAML rule: decision = %q, want deny", d)
	}
}

File: packages/agent-mesh/sdks/typescript/tests/policy.test.ts
import { PolicyEngine } from '../src/policy';
import { PolicyRule } from '../src/types';
import { writeFileSync, unlinkSync, mkdtempSync } from 'fs';
import { join } from 'path';
import { tmpdir } from 'os';

describe('PolicyEngine', () => {
  describe('evaluate()', () => {
    it('returns deny by default when no rules match', () => {
      const engine = new PolicyEngine();
      expect(engine.evaluate('some.action')).toBe('deny');
    });

    it('matches exact action names', () => {
      const engine = new PolicyEngine([
        { action: 'data.read', effect: 'allow' },
      ]);
      expect(engine.evaluate('data.read')).toBe('allow');
      expect(engine.evaluate('data.write')).toBe('deny');
    });

    it('matches wildcard (*) rules', () => {
      const engine = new PolicyEngine([
        { action: '*', effect: 'allow' },
      ]);
      expect(engine.evaluate('anything')).toBe('allow');
    });

    it('matches prefix wildcard (action.*)', () => {
      const engine = new PolicyEngine([
        { action: 'data.*', effect: 'allow' },
      ]);
      expect(engine.evaluate('data.read')).toBe('allow');
      expect(engine.evaluate('data.write')).toBe('allow');
      expect(engine.evaluate('data')).toBe('allow');
      expect(engine.evaluate('network.call')).toBe('deny');
    });

    it('evaluates rules in order (first match wins)', () => {
      const engine = new PolicyEngine([
        { action: 'data.delete', effect: 'deny' },
        { action: 'data.*', effect: 'allow' },
      ]);
      expect(engine.evaluate('data.delete')).toBe('deny');
      expect(engine.evaluate('data.read')).toBe('allow');
    });

    it('supports the review decision', () => {
      const engine = new PolicyEngine([
        { action: 'sensitive.op', effect: 'review' },
      ]);
      expect(engine.evaluate('sensitive.op')).toBe('review');
    });
  });

  describe('conditions', () => {
    it('matches when all conditions are met', () => {
      const engine = new PolicyEngine([
        {
          action: 'data.read',
          effect: 'allow',
          conditions: { role: 'admin' },
        },
      ]);
      expect(engine.evaluate('data.read', { role: 'admin' })).toBe('allow');
    });

    it('rejects when conditions are not met', () => {
      const engine = new PolicyEngine([
        {
          action: 'data.read',
          effect: 'allow',
          conditions: { role: 'admin' },
        },
      ]);
      expect(engine.evaluate('data.read', { role: 'user' })).toBe('deny');
    });

    it('handles multiple conditions', () => {
      const engine = new PolicyEngine([
        {
          action: 'data.write',
          effect: 'allow',
          conditions: { role: 'admin', env: 'production' },
        },
      ]);
      expect(engine.evaluate('data.write', { role: 'admin', env: 'production' })).toBe('allow');
      expect(engine.evaluate('data.write', { role: 'admin', env: 'staging' })).toBe('deny');
    });
  });

  describe('addRule()', () => {
    it('adds a rule that is evaluated', () => {
      const engine = new PolicyEngine();
      engine.addRule({ action: 'new.action', effect: 'allow' });
      expect(engine.evaluate('new.action')).toBe('allow');
    });

    it('appends to existing rules', () => {
      const engine = new PolicyEngine([
        { action: 'first', effect: 'deny' },
      ]);
      engine.addRule({ action: 'second', effect: 'allow' });
      expect(engine.getRules()).toHaveLength(2);
    });
  });

  describe('loadFromYAML()', () => {
    let tempDir: string;
    let yamlPath: string;

    beforeEach(() => {
      tempDir = mkdtempSync(join(tmpdir(), 'policy-test-'));
      yamlPath = join(tempDir, 'policy.yaml');
    });

    afterEach(() => {
      try { unlinkSync(yamlPath); } catch { /* ignore */ }
    });

    it('loads rules from a YAML file', async () => {
      writeFileSync(
        yamlPath,
        `rules:
  - action: "file.read"
    effect: "allow"
  - action: "file.delete"
    effect: "deny"
`,
      );

      const engine = new PolicyEngine();
      await engine.loadFromYAML(yamlPath);

      expect(engine.evaluate('file.read')).toBe('allow');
      expect(engine.evaluate('file.delete')).toBe('deny');
    });

    it('appends YAML rules to existing rules', async () => {
      writeFileSync(
        yamlPath,
        `rules:
  - action: "extra"
    effect: "review"
`,
      );

      const engine = new PolicyEngine([
        { action: 'existing', effect: 'allow' },
      ]);
      await engine.loadFromYAML(yamlPath);

      expect(engine.getRules()).toHaveLength(2);
      expect(engine.evaluate('extra')).toBe('review');
    });
  });
});
